src/persist-client/src/impl/machine.rs

```diff
@@ -0,0 +1,293 @@
+            .expect("clone_reader is infallible");
+        debug_assert_eq!(seqno, read_cap.seqno);
+        Ok(read_cap)
+    }
+
+    pub async fn append(
+        &mut self,
+        deadline: Instant,
+        writer_id: &WriterId,
+        keys: &[String],
+        desc: &Description<T>,
+    ) -> Result<Result<SeqNo, InvalidUsage>, LocationError> {
+        let res = self
+            .apply_unbatched_cmd(deadline, |_, state| state.append(writer_id, keys, desc))
+            .await?;
+        let (seqno, _) = match res {
+            Ok(x) => x,
+            Err(err) => return Ok(Err(err)),
+        };
+        Ok(Ok(seqno))
+    }
+
+    pub async fn compare_and_append(
+        &mut self,
+        deadline: Instant,
+        writer_id: &WriterId,
+        keys: &[String],
+        desc: &Description<T>,
+    ) -> Result<Result<SeqNo, InvalidUsage>, LocationError> {
+        let res = self
+            .apply_unbatched_cmd(deadline, |_, state| {
+                state.compare_and_append(writer_id, keys, desc)
+            })
+            .await?;
+        let (seqno, _) = match res {
+            Ok(x) => x,
+            Err(err) => return Ok(Err(err)),
+        };
+        Ok(Ok(seqno))
+    }
+
+    pub async fn downgrade_since(
+        &mut self,
+        deadline: Instant,
+        reader_id: &ReaderId,
+        new_since: &Antichain<T>,
+    ) -> Result<Result<SeqNo, InvalidUsage>, LocationError> {
+        let res = self
+            .apply_unbatched_cmd(deadline, |_, state| {
+                state.downgrade_since(reader_id, new_since)
+            })
+            .await?;
+        let (seqno, _) = match res {
+            Ok(x) => x,
+            Err(err) => return Ok(Err(err)),
+        };
+        Ok(Ok(seqno))
+    }
+
+    pub async fn expire_writer(
+        &mut self,
+        deadline: Instant,
+        writer_id: &WriterId,
+    ) -> Result<SeqNo, LocationError> {
+        let res = self
+            .apply_unbatched_cmd(deadline, |seqno, state| {
+                state.expire_writer(seqno, writer_id)
+            })
+            .await?;
+        let seqno = match res {
+            Ok((seqno, ())) => seqno,
+            Err(NoOp { seqno }) => seqno,
+        };
+        Ok(seqno)
+    }
+
+    pub async fn expire_reader(
+        &mut self,
+        deadline: Instant,
+        reader_id: &ReaderId,
+    ) -> Result<SeqNo, LocationError> {
+        let res = self
+            .apply_unbatched_cmd(deadline, |seqno, state| {
+                state.expire_reader(seqno, reader_id)
+            })
+            .await?;
+        let seqno = match res {
+            Ok((seqno, ())) => seqno,
+            Err(NoOp { seqno }) => seqno,
+        };
+        Ok(seqno)
+    }
+
+    pub async fn snapshot(
+        &mut self,
+        deadline: Instant,
+        as_of: &Antichain<T>,
+    ) -> Result<Result<Vec<String>, InvalidUsage>, LocationError> {
+        // TODO: This fetches the latest state and uses that to determine if we
```

Comment:
hmmm could you say a tiny bit more about this? My understanding is that our local `since` will hold back compaction, so as long as the requested as_of is >= our read capability's since we should be good to go. what am i missing?

Reply:
I _think_ Dan was talking about this:
 - right now, every `snapshot` (and `next_listen_batch()`) does a consensus read, even though they might not have to because our locally available state would be sufficient to serve the request
 - consensus reads are expensive
 - it's mentioned below: it would be nice if we could passively listen to changes from Consensus

Reply:
Specifically I'm responding to:

```
It's possible that this `as_of` could be served by the version we happen to have cached, but not by this one we fetch.
```

This is confusing to me because the reader's `since` frontier should hold back compaction to `since`, and my understanding is that the reader has perfect knowledge of its `since` frontier unless the readers lease expires.

Obviously, if the lease expires the reader can't service this request but in that case the `as_of` is irrelevant. The way I interpreted Dan's comment is "we might be able to serve this as_of based on what we know locally, but some compaction might advance the collection's since frontier, and make it impossible for us to serve this request with a more up to date state", which conflicts with my understanding of the reader's since frontier.